src/dune_engine/sandbox.ml

```diff
@@ -41,26 +41,39 @@ type t =
 let dir t = t.dir
 let map_path t p = Path.Build.append t.dir p
 
-let rec copy_recursively (src_kind : Unix.file_kind) ~src ~dst =
-  match src_kind with
-  | S_REG -> Io.copy_file ~src ~dst ()
-  | S_DIR ->
-    (match Path.Untracked.readdir_unsorted_with_kinds src with
-     | Error e -> Unix_error.Detailed.raise e
-     | Ok contents ->
-       Path.mkdir_p dst;
-       List.iter contents ~f:(fun (name, kind) ->
-         copy_recursively kind ~src:(Path.relative src name) ~dst:(Path.relative dst name)))
-  | _ ->
-    User_error.raise
-      ~hints:
-        [ Pp.text "re-run dune to delete the stale artifact, or manually delete this file"
+let copy_recursively =
+  let chmod = Path.Permissions.add Path.Permissions.write in
+  let rec loop what ~src ~dst =
+    match what with
+    | `File -> Io.copy_file ~chmod ~src ~dst ()
+    | `Dir perms ->
+      (match Path.Untracked.readdir_unsorted_with_kinds src with
+       | Error e -> Unix_error.Detailed.raise e
+       | Ok contents ->
+         let perms = chmod perms in
+         Path.mkdir_p ~perms dst;
```

Comment:
@snowleopard this is the line that makes it appear as if we're losing permissions. But we're actually just copying them accurately. Previously, we would just do `Path.mkdir_p` which always created things as `755`.

I could move this into a separate PR if you prefer, but to me it seemed like a similar enough of a bug (incorrect permissions) to just stuff it here.

Reply:
Ah, I see, thanks.

Reply:
@snowleopard actually, should we also add executable permissions to the directory? Without executable permissions, it's not possible to write anything to the directory anyway.

Reply:
Yeah, we might as well do that. I guess it doesn't matter much in practice because unreadable directories just don't show up often.